glib/Idle.cs

```diff
@@ -44,44 +50,84 @@ protected override bool Invoke ()
 				return real_handler ();
 			}
 		}
-		
+
 		private Idle ()
 		{
 		}
-		
-		[DllImport("libglib-2.0-0.dll", CallingConvention=CallingConvention.Cdecl)]
-		static extern uint g_idle_add (SourceProxy.GSourceFuncInternal d, IntPtr data);
+
+		const int G_PRIORITY_DEFAULT_IDLE = 200;
+
+		[DllImport ("libglib-2.0-0.dll", CallingConvention = CallingConvention.Cdecl)]
+		static extern uint g_idle_add_full (int priority, SourceProxy.GSourceFuncInternal d, IntPtr data, DestroyNotify notify);
 
 		public static uint Add (IdleHandler hndlr)
 		{
 			IdleProxy p = new IdleProxy (hndlr);
-			p.ID = g_idle_add (SourceProxy.SourceHandler, (IntPtr)p.handle);
-			Source.Add (p);
+
+			p.ID = g_idle_add_full (G_PRIORITY_DEFAULT_IDLE, SourceProxy.SourceHandler, (IntPtr)p.handle, NotifyHandler);
+
+			lock (idle_handlers) {
+				if (p.needsAdd) {
+					p.needsAdd = false;
+					idle_handlers [p.ID] = p;
+				}
+			}
 
 			return p.ID;
 		}
-		
-		[DllImport("libglib-2.0-0.dll", CallingConvention=CallingConvention.Cdecl)]
+
+		internal static Dictionary<uint, IdleProxy> idle_handlers = new Dictionary<uint, IdleProxy> ();
+
+		[DllImport ("libglib-2.0-0.dll", CallingConvention = CallingConvention.Cdecl)]
 		static extern bool g_source_remove (uint id);
-                                                                                
+
 		public static bool Remove (IdleHandler hndlr)
 		{
 			bool result = false;
 
-			lock (Source.source_handlers) {
-				foreach (KeyValuePair<uint, SourceProxy> kvp in Source.source_handlers) {
+			lock (idle_handlers) {
+				foreach (KeyValuePair<uint, IdleProxy> kvp in idle_handlers.ToArray ()) {
 					uint code = kvp.Key;
-					IdleProxy p = kvp.Value as IdleProxy;
-				
+					IdleProxy p = kvp.Value;
+
 					if (p != null && p.real_handler == hndlr) {
 						result = g_source_remove (code);
-						p.Remove ();
 					}
 				}
 			}
 
 			return result;
 		}
+
+		#region IdleProxy DestroyNotify
+		static void ReleaseGCHandle (IntPtr data)
+		{
+			if (data == IntPtr.Zero)
+				return;
+
+			GCHandle gch = (GCHandle)data;
+			IdleProxy proxy = (IdleProxy)gch.Target;
+
+			lock (idle_handlers) {
+				if (proxy.needsAdd)
+					proxy.needsAdd = false;
+				else
+					idle_handlers.Remove (proxy.ID);
+			}
+
+			gch.Free ();
```

Comment:
My concern here is that the `IdleProxy` object still has a reference to the GCHandle and we run the risk of accidentally double-freeing the handle now. Could we make `SourceProxy` IDisposable, then change this code to call `Dispose` on it and null out the GCHandle field as soon as it's freed?